feat(common): new pipe to support proper case conversion

AlmogShaul opened the pull request:
Please check if the PR fulfills these requirements

The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
Tests for the changes have been added (for bug fixes / features)
Docs have been added / updated (for bug fixes / features)
What kind of change does this PR introduce? (check one with "x")

[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
What is the current behavior? (You can also link to an open issue here)

What is the new behavior?

Does this PR introduce a breaking change? (check one with "x")

[ ] Yes
[x] No
If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

Other information:

Add propercase_pipe which convert string to propercase

'hello there'->'Hello there'

AlmogShaul committed:
# This is a combination of 8 commits.
# The first commit's message is:
Proper case pipe

# This is the 2nd commit message:

Proper case pipe

# This is the 3rd commit message:

Proper case pipe

# This is the 4th commit message:

Proper case pipe

# This is the 5th commit message:

# This is a combination of 2 commits.
# The first commit's message is:

Proper case pipe

# This is the 2nd commit message:

Proper case pipe

# This is the 6th commit message:

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

# This is the 7th commit message:

# This is a combination of 3 commits.
# The first commit's message is:
Proper case pipe

# This is the 2nd commit message:

Proper case pipe

# This is the 3rd commit message:

Proper case pipe

# This is the 8th commit message:

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

AlmogShaul committed:
Proper case pipe

AlmogShaul committed:
# This is a combination of 3 commits.
# The first commit's message is:
Proper case pipe

# This is the 2nd commit message:

Proper case pipe

# This is the 3rd commit message:

Proper case pipe

AlmogShaul committed:
Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

test: reactivate the remaining disabled tests in Edge (#11188)

Fixes #4756

style(dom_renderer): use const (#11229)

build: update the symlinks scripts for Windows to new packaging (#11192)

docs(CHANGELOG): fix provider syntax (#11258)

fixes #11256

fix typo (#11265)

meesage => message

refactor(EventManager): remove ListWrapper (#11363)

fix(compiler): correctly type event handler proxy functions

fix(compiler): error when `NgModule.bootstrap` contains `undefined` or `null`

fix(lazy-loading): fix an issue with webpack and lazy loader. (#11387)

The issue was introduced in PR#11049.

feat(benchmarks): add targetable benchmarks back

refactor(benchmarks): align tree benchmark with largetable benchmark

- add ng2_switch benchmark to track `ngFor` over `ngSwitch`
- measure create only, createDestroy and update
- simplify the created dom
- always add a style binding

ci: update SauceLabs badge when running CI on master (#11352)

build(test.sh): clear dist directory when the script starts

This is to prevent stale files causing the tests to fail when we run them locally after checking out a new revision.

build: remove obsolete presubmit.sh script

test: cleanup playground/src/bootstrap.ts file

docs: misc fixes.

docs(common_module): Fix macro format
docs(number_pipe): Add missing period sign
docs(date_pipe): Fix suffix consistency
docs(date_pipe): Fix missing quote
docs(number_pipe): Fix incorrect article

Looks like the word "Polyfill" does not start with a vowel pronunciation.

docs(location_strategy): Fix code format

Add missing \`\`\` at start.

docs(i18n_plural_pipe): Fix code format
docs(location): Add missing period sign
refactor(ngSwitch): fix typo on parameter
docs(di): Add missing quote
docs(compiler): Fix typo
docs(compiler): Add missing period sign
docs(directives): Fix description for styles parameter
docs(location_strategy): Add code language

Revert for misunderstanding.

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

test: reactivate the remaining disabled tests in Edge (#11188)

Fixes #4756

style(dom_renderer): use const (#11229)

build: update the symlinks scripts for Windows to new packaging (#11192)

docs(CHANGELOG): fix provider syntax (#11258)

fixes #11256

fix typo (#11265)

meesage => message

refactor(EventManager): remove ListWrapper (#11363)

fix(compiler): correctly type event handler proxy functions

fix(compiler): error when `NgModule.bootstrap` contains `undefined` or `null`

fix(lazy-loading): fix an issue with webpack and lazy loader. (#11387)

The issue was introduced in PR#11049.

feat(benchmarks): add targetable benchmarks back

refactor(benchmarks): align tree benchmark with largetable benchmark

- add ng2_switch benchmark to track `ngFor` over `ngSwitch`
- measure create only, createDestroy and update
- simplify the created dom
- always add a style binding

ci: update SauceLabs badge when running CI on master (#11352)

build(test.sh): clear dist directory when the script starts

This is to prevent stale files causing the tests to fail when we run them locally after checking out a new revision.

build: remove obsolete presubmit.sh script

test: cleanup playground/src/bootstrap.ts file

docs: misc fixes.

docs(common_module): Fix macro format
docs(number_pipe): Add missing period sign
docs(date_pipe): Fix suffix consistency
docs(date_pipe): Fix missing quote
docs(number_pipe): Fix incorrect article

Looks like the word "Polyfill" does not start with a vowel pronunciation.

docs(location_strategy): Fix code format

Add missing \`\`\` at start.

docs(i18n_plural_pipe): Fix code format
docs(location): Add missing period sign
refactor(ngSwitch): fix typo on parameter
docs(di): Add missing quote
docs(compiler): Fix typo
docs(compiler): Add missing period sign
docs(directives): Fix description for styles parameter
docs(location_strategy): Add code language

Revert for misunderstanding.

AlmogShaul committed:
# This is a combination of 2 commits.
# The first commit's message is:

Proper case pipe

# This is the 2nd commit message:

Proper case pipe

AlmogShaul committed:
Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

AlmogShaul committed:
Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

AlmogShaul committed:
Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

AlmogShaul committed:
# This is a combination of 2 commits.
# The first commit's message is:

Proper case pipe

# This is the 2nd commit message:

Proper case pipe

AlmogShaul committed:
Proper case pipe

AlmogShaul committed:
Proper case pipe

vicb commented:
- Please merge the commits,
- I think it would make sense to move all case conversion pipes in a single file

marclaval committed:
test: reactivate the remaining disabled tests in Edge (#11188)

Fixes #4756

style(dom_renderer): use const (#11229)

build: update the symlinks scripts for Windows to new packaging (#11192)

docs(CHANGELOG): fix provider syntax (#11258)

fixes #11256

fix typo (#11265)

meesage => message

refactor(EventManager): remove ListWrapper (#11363)

fix(compiler): correctly type event handler proxy functions

fix(compiler): error when `NgModule.bootstrap` contains `undefined` or `null`

fix(lazy-loading): fix an issue with webpack and lazy loader. (#11387)

The issue was introduced in PR#11049.

feat(benchmarks): add targetable benchmarks back

refactor(benchmarks): align tree benchmark with largetable benchmark

- add ng2_switch benchmark to track `ngFor` over `ngSwitch`
- measure create only, createDestroy and update
- simplify the created dom
- always add a style binding

ci: update SauceLabs badge when running CI on master (#11352)

build(test.sh): clear dist directory when the script starts

This is to prevent stale files causing the tests to fail when we run them locally after checking out a new revision.

build: remove obsolete presubmit.sh script

test: cleanup playground/src/bootstrap.ts file

docs: misc fixes.

docs(common_module): Fix macro format
docs(number_pipe): Add missing period sign
docs(date_pipe): Fix suffix consistency
docs(date_pipe): Fix missing quote
docs(number_pipe): Fix incorrect article

Looks like the word "Polyfill" does not start with a vowel pronunciation.

docs(location_strategy): Fix code format

Add missing \`\`\` at start.

docs(i18n_plural_pipe): Fix code format
docs(location): Add missing period sign
refactor(ngSwitch): fix typo on parameter
docs(di): Add missing quote
docs(compiler): Fix typo
docs(compiler): Add missing period sign
docs(directives): Fix description for styles parameter
docs(location_strategy): Add code language

Revert for misunderstanding.

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

Proper case pipe

test: reactivate the remaining disabled tests in Edge (#11188)

Fixes #4756

style(dom_renderer): use const (#11229)

build: update the symlinks scripts for Windows to new packaging (#11192)

docs(CHANGELOG): fix provider syntax (#11258)

fixes #11256

fix typo (#11265)

meesage => message

refactor(EventManager): remove ListWrapper (#11363)

fix(compiler): correctly type event handler proxy functions

fix(compiler): error when `NgModule.bootstrap` contains `undefined` or `null`

fix(lazy-loading): fix an issue with webpack and lazy loader. (#11387)

The issue was introduced in PR#11049.

feat(benchmarks): add targetable benchmarks back

refactor(benchmarks): align tree benchmark with largetable benchmark

- add ng2_switch benchmark to track `ngFor` over `ngSwitch`
- measure create only, createDestroy and update
- simplify the created dom
- always add a style binding

ci: update SauceLabs badge when running CI on master (#11352)

build(test.sh): clear dist directory when the script starts

This is to prevent stale files causing the tests to fail when we run them locally after checking out a new revision.

build: remove obsolete presubmit.sh script

test: cleanup playground/src/bootstrap.ts file

docs: misc fixes.

docs(common_module): Fix macro format
docs(number_pipe): Add missing period sign
docs(date_pipe): Fix suffix consistency
docs(date_pipe): Fix missing quote
docs(number_pipe): Fix incorrect article

Looks like the word "Polyfill" does not start with a vowel pronunciation.

docs(location_strategy): Fix code format

Add missing \`\`\` at start.

docs(i18n_plural_pipe): Fix code format
docs(location): Add missing period sign
refactor(ngSwitch): fix typo on parameter
docs(di): Add missing quote
docs(compiler): Fix typo
docs(compiler): Add missing period sign
docs(directives): Fix description for styles parameter
docs(location_strategy): Add code language

Revert for misunderstanding.

metanav commented on modules/@angular/common/src/pipes/propercase_pipe.ts:
```diff
@@ -0,0 +1,31 @@
+    if (!isString(value)) {
+      throw new InvalidPipeArgumentError(ProperCasePipe, value);
+    }
+    return value[0].toUpperCase() + value.substr(1);
```
I think 'value' should be converted to lowercase first to cover all cases:
 'hello there or HELLO THERE or hElLo tHeRe' will be transformed to 'Hello there'.

value = value.toLowerCase();
return value[0].toUpperCase() + value.substr(1);

AlmogShaul replied:
right,I will fix it

AlmogShaul replied:
@vicb To make a single file with all the case conversion pipes, or to wait with that?

AlmogShaul committed:
Proper case pipe

AlmogShaul commented:
@metanav I need help squashing my pushed commits.

vicb commented:
1 - merge all case conversion pipes,
2 - squash the commits,
3 - the resulting commit comment should stick the conventions: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format 

To squash commits:
```
git checkout master
git fetch
git merge origin/master --ff

git checkout camelCasePipe
git rebase master
git rebase matser -i  # pick the first commit, and replace pick with squash for following ones
```

(you'll find plenty of help in github... or in google)

AlmogShaul commented:
I authored these commits

AlmogShaul committed:
Revert "test: reactivate the remaining disabled tests in Edge (#11188)"

This reverts commit 7b2f95c3a228f311800bdde6b9ea42cc4f3f2590.
